feat: add remaining pyarrow types

aaravind100 opened the pull request:
Add the remaining pyarrow types as per the conversation in #1676.

Below types are added:

- null
- date32
- date64
- duration
- float16
- time32
- time64
- map_
- binary
- large_binary
- large_string

Note: `null` requires field `nullable` option to be set to `True` for successful validation.

aaravind100 commented:
@cosmicBboy any suggestions for fixing this test?

[FAILED tests/core/test_pandas_engine.py::test_pandas_data_type_coerce[ArrowMap] - assert 0 > 0](https://github.com/unionai-oss/pandera/actions/runs/9723147115/job/26837913815?pr=1720#step:7:2561)

It does appear to raise a `ParserError`, but the count is 0.

cosmicBboy commented:
@aaravind100 it looks like the underlying error is:
```
NotImplementedError: Converting strings to map<int64, int64> is not implemented
```

It looks like this test case needs to be updated to contain a list of dicts where keys are ints and values are strings? https://github.com/unionai-oss/pandera/pull/1720/files#diff-358f62ffd0dea15c0dcd1efc34883303f1d085f35ebb8d3923c781872f6eb563R317-R325

aaravind100 commented:
@cosmicBboy implementing a coerce value method fixed it. It was not able to get the failures cases.

aaravind100 commented:
Rebased and fixed all the failing tests :smile:

There was this weird case where pandas 2.0.3 was sneaking in the nox `"tests(extra='core', pydantic='2.3.0', python='3.8', pandas='1.5.3')"` session.

https://github.com/unionai-oss/pandera/actions/runs/9735487403/job/26864752599?pr=1720#step:7:133

cosmicBboy commented:
thanks @aaravind100 !
